### ISSUE TITLE
Can I use TF Privacy on Browser or Edge Devices?

### ISSUE THREAD

**Issue by Bernardo-Favoreto** [tensorflow/privacy#93]
I was wondering if I could use the TF privacy API on a browser or edge device, e.g. smartphone. 

My goal is to train a model in Federated Learning _with_ differential privacy. This means that the DP part should be on the user's end, so that the server only receives differentially private data (the gradients).

Is that possible? Is there an easier way to achieve what I'm trying?

Thanks!

**Comment by galenmandrew** [comment]
You are describing what is called "local differential privacy", where even the server only sees the noised gradients. The public classes currently implemented in tensorflow_privacy are designed for "global differential privacy" in which a trusted aggregator averages the unnoised gradients and then adds noise before publishing the result.

Local DP is certainly possible within tensorflow_privacy. You would need to write a DPQuery (e.g., similar to GaussianAverageQuery) which adds the noise as part of "preprocess_record" and then "get_noised_result" would simply average the already-noised records.

However, you should be aware that in local DP a much higher level of noise is required for the same privacy guarantee, so the privacy/utility tradeoff is much worse.
